### PR TITLE
Load library images through storage

### PR DESCRIPTION
**Pull request by cwillisf**
### Resolves

Part of #4598

### Proposed Changes

This PR adds support for loading libraries through `scratch-storage`. It does this with a new class, `ScratchImage`, which supports loading images either by URI or by storage info (asset type + asset ID).

### Reason for Changes

This allows the libraries to work without an Internet connection, such as in the case of Scratch Desktop.

### Test Coverage

Existing library tests have been updated for these changes.

### Browser Coverage
Check the OS/browser combinations tested (At least 2)

Mac
 * [x] Chrome 
 * [ ] Firefox 
 * [x] Safari
 
Windows
 * [x] Chrome 
 * [x] Firefox
 * [x] Edge
 
Chromebook
 * [ ] Chrome
 
iPad
* [ ] Safari

Android Tablet
* [ ] Chrome

### CODE REVIEW

**Comment by paulkaplan** [comment]
@cwillisf i put up https://llk.github.io/scratch-gui/library-storage/ so we could test performance without worrying about dev environment. One thing that jumps out immediately is comparing the performance of opening the costume editor the second time, after all the images have loaded. On /develop, after clicking the costume library button you see the modal in about 500ms. On that branch, it takes 1.5-2s to see the library. I'm going to keep investigating other situations, but that one seems like a blocker.

File: src/components/library/library.jsx [hunk 1]
```diff
@@ -28,6 +29,47 @@ const messages = defineMessages({
...
+ * Find the AssetType which corresponds to a particular file extension. For example, 'png' => AssetType.ImageBitmap.
+ * @param {string} fileExtension - the file extension to look up.
+ * @returns {AssetType} - the AssetType corresponding to the extension, if any.
+ */
+const getAssetTypeForFileExtension = function (fileExtension) {
+    const compareOptions = {
+        sensitivity: 'accent',
+        usage: 'search'
+    };
+    for (const assetTypeId of Object.keys(storage.AssetType)) {
```
[the remark this answers is not in the document]
Comment (paulkaplan) [reply src/components/library/library.jsx]: I think you just want `for ... in ...` instead of for of

**Comment by paulkaplan** [comment]
@cwillisf i think the fact that it keeps making network requests long after you've closed the modal seems kinda bad.

File: src/components/library-item/library-item.jsx [hunk 2]
```diff
@@ -35,10 +36,12 @@ class LibraryItemComponent extends React.PureComponent {
                             />
                         </div>
                     ) : null}
-                    <img
-                        className={styles.featuredImage}
-                        src={this.props.iconURL}
-                    />
+                    {this.props.iconSource ? (
```
[the remark this answers is not in the document]
Comment (cwillisf) [reply src/components/library-item/library-item.jsx]: While I agree that your proposal looks cleaner, the `condition ? element : null` pattern shows up throughout this file. I feel like I should go for consistency and we can look at refactoring later...?

**Comment by chrisgarrity** [comment]
If there was a way to prioritize the other (invisible) costumes of a sprite that is visible, that would make the animation even more responsive. Right now there's a bit of a delay between mousing over the sprite and the animation of the costumes. However, it's already better than what's on production for a slow connection, so no reason to hold it up for that.

**Comment by cwillisf** [comment]
I'm going to hold off on merging this for another week to reduce our testing burden on April 4. I'll merge this after the testing session.

LLK/scratch-storage#83 has been merged, though, which should reduce the potential for negative performance impact from this change.

**Pull request merged**